### DISCUSSION TITLE
Announcement: debugWIRE hardware debugger ready to be used

### DISCUSSION THREAD

**Discussion by felias-fogg** [SpenceKonde/ATTinyCore#643] [Show and tell]
Hi, this not exactly an issue. So no work involved! 

I just wanted to announce that my Arduino-based debugWIRE hardware debugger [dw-link](https://github.com/felias-fogg/dw-link) is ready and has been tested on all the MCUs supported by ATTinyCore. You just need an Uno (or equivalent) and 6 jumper wires, you have to install avr-gdb, and off you go.

Best,
Bernhard

**Comment by SpenceKonde** [comment]
Goddamn! Now that is pretty damned cool. Good to be aware something like this exists - though having never used a hardware debugger, I suspect that the learning curve on avr-gdb would make it hard to justify using. I can think of a very small number ofproblems over the years that made me wish I had a hardware debug mechanism (the issue with tinySoftwareSerial's available/read functions not working for sketches with no local variables comes to mind, though I eventually figured out a workaround an the essential nature ofthe bug by doing file comparison on the assembly, though it was tricky because the generated code was completely reorganized. It was one of those tiny changes that completely shifted the way the compiler organized everything. Yeah something like this would have been very helpful there.... (the problem, apparently, was that the stack wasn't getting properly initialized unless somewhere in the sketch and core, a local variable was used o_o wacky... 

Moved to discussion because the issues section is my list of action items.

**Reply by prandeamus** [reply]
You might also be interested in the article written here by felias. https://hinterm-ziel.de/index.php/2022/01/13/a-debugwire-hardware-debugger-for-less-than-10-e/

**Comment by felias-fogg** [comment]
Hi Spence,

this sounds like a bug from the collection of nightmare creatures! And although debugging with GDB beats printf-debugging anytime, I guess even with GDB you would have had a hard time locating this bug. In any case, most of the time I use it to find the bugs I myself have inserted into the code. Just remember the quote "If debugging is the process of removing software bugs, then programming must be the process of putting them in” by Dijkstra.
